### PR TITLE
{Core} Add `--allow-preview` to `az extension add/update` and `az upgrade`

### PR DESCRIPTION
**Pull request by AllyW**
**Related command**

az extension add/update
az upgrade

**Description**

**Testing Guide**

**History Notes**

[Extension] `az extension add/update`: Add `--allow-preview` to distinguish stable-only installation with preview-included installation
[Upgrade] `az upgrade`: Add `--allow-preview` to distinguish stable-only extension installation with preview-included extension installation

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

File: src/azure-cli-core/azure/cli/core/extension/_resolve.py [hunk 1]
```diff
@@ -86,6 +95,7 @@ def resolve_from_index(extension_name, cur_version=None, index_url=None, target_
         index_url (str, optional):  Defaults to None.
         target_version (str, optional): Version of extension to install. Defaults to latest version.
         cli_ctx (, optional): CLI Context. Defaults to None.
+        allow_preview(bool, optional): Flag to allow installing preview extensions. Default to None.
```
Comment (bebound) [inline src/azure-cli-core/azure/cli/core/extension/_resolve.py:98]: It's better to default a boolean to a boolean value.

File: src/azure-cli-core/azure/cli/core/extension/_resolve.py [hunk 2]
```diff
@@ -77,7 +77,8 @@ def _get_version_compatibility_feedback(candidates: List[dict]) -> str:
         return e.args[0]
 
 
-def resolve_from_index(extension_name, cur_version=None, index_url=None, target_version=None, cli_ctx=None):
+def resolve_from_index(extension_name, cur_version=None, index_url=None, target_version=None, cli_ctx=None,
```
Comment (kairu-ms) [inline src/azure-cli-core/azure/cli/core/extension/_resolve.py:80]: Will this function be called in `az upgrade`?

**Review by kairu-ms — APPROVED** [review]

File: src/azure-cli-core/azure/cli/core/extension/__init__.py [hunk 3]
```diff
@@ -374,3 +374,9 @@ def is_preview_from_semantic_version(version):
     from packaging.version import parse
     parsed_version = parse(version)
     return bool(parsed_version.pre and parsed_version.pre[0] in ["a", "b"])
+
+
+def is_stable_from_metadata(item):
+    return not (item["metadata"].get(EXT_METADATA_ISPREVIEW, False) or
+                item["metadata"].get(EXT_METADATA_ISEXPERIMENTAL, False) or
+                is_preview_from_semantic_version(item["metadata"]['version']))
```
Comment (zhoxing-ms) [inline src/azure-cli-core/azure/cli/core/extension/__init__.py:382]:
```suggestion
    return item["metadata"].get(EXT_METADATA_ISPREVIEW, False) and
                item["metadata"].get(EXT_METADATA_ISEXPERIMENTAL, False) and
                not is_preview_from_semantic_version(item["metadata"]['version']
```
May I ask could the above code be expressed more concisely?

Comment (zhoxing-ms) [reply src/azure-cli-core/azure/cli/core/extension/__init__.py:382]: Also, I'd like to remind that some old extensions may have already been GA to a stable version, but they may have forgotten to change the `azext.isPreview`.
Maybe we can either correct them or simply ignore the `azext.isPreview` if it doesn't have any practical effect

Comment (evelyn-ys) [reply src/azure-cli-core/azure/cli/core/extension/__init__.py:382]: does `item` always have value and contains `metadata` key?

File: src/azure-cli/azure/cli/command_modules/util/_params.py [hunk 4]
```diff
@@ -38,6 +38,7 @@ def load_arguments(self, _):
     with self.argument_context('upgrade') as c:
         c.argument('update_all', options_list=['--all'], arg_type=get_three_state_flag(), help='Enable updating extensions as well.', default='true')
         c.argument('yes', options_list=['--yes', '-y'], action='store_true', help='Do not prompt for checking release notes.')
+        c.argument('allow_preview', options_list=['--allow-preview-extensions', '--allow-preview'], help="Include preview packages for extension installation, if exists")
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/util/_params.py:41]: Consider using `arg_type=get_three_state_flag()`.

**Comment by jiasli** [comment]
> [Util] `az upgrade`: Add `--allow-preview`

This makes `az upgrade` be under `Util` section. In https://learn.microsoft.com/en-us/cli/azure/release-notes-azure-cli, `az upgrade` sometimes appears under `Packaging` section, while sometimes it appears under `Upgrade` section. We should unify it. @bebound

**Comment by jiasli** [comment]
> [Extension] `az extension add/update`: Add `--allow-preview`

It's better to elaborate what `--allow-preview` does.

**Comment by bebound** [comment]
> > [Util] `az upgrade`: Add `--allow-preview`
> 
> This makes `az upgrade` be under `Util` section. In https://learn.microsoft.com/en-us/cli/azure/release-notes-azure-cli, `az upgrade` sometimes appears under `Packaging` section, while sometimes it appears under `Upgrade` section. We should unify it. @bebound

Let's use `Upgrade`, it's introduced in https://learn.microsoft.com/en-us/cli/azure/release-notes-azure-cli#upgrade-2

**Review by bebound — DISMISSED** [review]

**Comment by evelyn-ys** [comment]
> [Extension] az extension add/update: Add --allow-preview
> [Util] az upgrade: Add --allow-preview

Both are under **Description** section, should they be moved to **History Notes** section?

File: src/azure-cli-core/azure/cli/core/extension/__init__.py (hunk 3 above)
Comment (AllyW) [reply src/azure-cli-core/azure/cli/core/extension/__init__.py:382]: > does `item` always have value and always contain `metadata` key?

Yes, it's a property defined in class Extension and initially is an empty dict.

Comment (AllyW) [reply src/azure-cli-core/azure/cli/core/extension/__init__.py:382]: > In addition, I'd like to remind that some old extensions may have already been GA to a stable version, but they may have forgotten to change the `azext.isPreview`. Maybe we can either correct them, or simply ignore the `azext.isPreview` if it doesn't have any practical effect

yes. For extension packages, we plan to correct their package versions and version tags in future pipelines.

File: src/azure-cli/azure/cli/command_modules/extension/__init__.py [hunk 5]
```diff
@@ -66,6 +66,7 @@ def load_arguments(self, command):
                        help='Space-separated list of extra URLs of package indexes to use. This should point to a repository compliant with PEP 503 (the simple repository API) or a local directory laid out in the same format.', arg_group='Experimental Pip')
             c.ignore('_subscription')  # hide global subscription param
             c.argument('system', action='store_true')
+            c.argument('allow_preview', options_list=['--allow-preview'], help="Include preview packages for extension installation, if exists")
```
[the remark this answers is not in the document]
Comment (AllyW) [reply src/azure-cli/azure/cli/command_modules/extension/__init__.py]: For users, preview or stable is what they might pay attention to. Detailed format for preview extension packages is required to be noticed for extension module developers, in my view.

**Review by evelyn-ys — APPROVED** [review]

**Review by bebound — APPROVED** [review]

**Review by kairu-ms — APPROVED** [review]

**Comment by bebound** [comment]
@AllyW This PR breaks CI test with:
```
2023-12-13T03:38:57.0958900Z ==================================== ERRORS ====================================
2023-12-13T03:38:57.0959215Z ___________________ ERROR collecting tests/test_extension.py ___________________
2023-12-13T03:38:57.0959952Z ImportError while importing test module '/usr/lib64/az/lib/python3.9/site-packages/azure/cli/core/tests/test_extension.py'.
2023-12-13T03:38:57.0960285Z Hint: make sure your test modules/packages have valid Python names.
2023-12-13T03:38:57.0960501Z Traceback:
2023-12-13T03:38:57.0960738Z /usr/lib/python3.9/importlib/__init__.py:127: in import_module
2023-12-13T03:38:57.0961007Z     return _bootstrap._gcd_import(name[level:], package, level)
2023-12-13T03:38:57.0961444Z /usr/lib64/az/lib/python3.9/site-packages/azure/cli/core/tests/test_extension.py:20: in <module>
2023-12-13T03:38:57.0961968Z     from azure.cli.core.extension.tests.latest import IndexPatch, mock_ext
2023-12-13T03:38:57.0962377Z E   ModuleNotFoundError: No module named 'azure.cli.core.extension.tests'
2023-12-13T03:38:57.0962782Z -------- generated xml file: /azure_cli_test_result/azure-cli-core.xml ---------
2023-12-13T03:38:57.0963093Z =========================== short test summary info ============================
2023-12-13T03:38:57.0963346Z ERROR tests/test_extension.py
2023-12-13T03:38:57.0963600Z !!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
```
https://dev.azure.com/azclitools/public/_build/results?buildId=114970&view=logs&j=b67de874-13d2-5286-c044-deb7a6e6a9ee&t=f22bcdb8-df09-5097-beed-939fe0350d62

I guess you can follow https://github.com/Azure/azure-cli/pull/13572 to fix this temporally.

**Comment by AllyW** [comment]
> @AllyW This PR breaks CI test with:
> 
> ```
> 2023-12-13T03:38:57.0958900Z ==================================== ERRORS ====================================
> 2023-12-13T03:38:57.0959215Z ___________________ ERROR collecting tests/test_extension.py ___________________
> 2023-12-13T03:38:57.0959952Z ImportError while importing test module '/usr/lib64/az/lib/python3.9/site-packages/azure/cli/core/tests/test_extension.py'.
> 2023-12-13T03:38:57.0960285Z Hint: make sure your test modules/packages have valid Python names.
> 2023-12-13T03:38:57.0960501Z Traceback:
> 2023-12-13T03:38:57.0960738Z /usr/lib/python3.9/importlib/__init__.py:127: in import_module
> 2023-12-13T03:38:57.0961007Z     return _bootstrap._gcd_import(name[level:], package, level)
> 2023-12-13T03:38:57.0961444Z /usr/lib64/az/lib/python3.9/site-packages/azure/cli/core/tests/test_extension.py:20: in <module>
> 2023-12-13T03:38:57.0961968Z     from azure.cli.core.extension.tests.latest import IndexPatch, mock_ext
> 2023-12-13T03:38:57.0962377Z E   ModuleNotFoundError: No module named 'azure.cli.core.extension.tests'
> 2023-12-13T03:38:57.0962782Z -------- generated xml file: /azure_cli_test_result/azure-cli-core.xml ---------
> 2023-12-13T03:38:57.0963093Z =========================== short test summary info ============================
> 2023-12-13T03:38:57.0963346Z ERROR tests/test_extension.py
> 2023-12-13T03:38:57.0963600Z !!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
> ```
> 
> https://dev.azure.com/azclitools/public/_build/results?buildId=114970&view=logs&j=b67de874-13d2-5286-c044-deb7a6e6a9ee&t=f22bcdb8-df09-5097-beed-939fe0350d62
> 
> It's similar to #13572. I'm investigating it.

Test adjust pr: https://github.com/Azure/azure-cli/pull/28025

**Pull request merged**